sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-            
-        condensed_a_data[tv].extend(episodes)
-        condensed_a_data[tv] = sorted(condensed_a_data[tv], key=lambda item: item[0])
-        #print(len(condensed_a_data[tv]))
-    #condensed_a_data = {k: v for k, v in sorted(condensed_a_data.items(), key=lambda item: item[1][0][0])}
-    #print([(k,[x[0] for x in v]) for k,v in condensed_a_data.items()])
-    print('Attacker hosts: ', (set([x.split('-')[1] for x in condensed_a_data.keys()])))
-    return (condensed_a_data, condensed_v_data)
-
-## translate technical nodes to human readable
-def translate(label, root=False):
+
+# Step 6.4: Group episodes (state sequences) per (team, victim) pair
+def group_episodes_per_av(state_sequences):
+    # Experiment: attack graph for one victim w.r.t time
+    victim_episodes = dict()  # Episodes per (team, victim)
+    for attack, episodes in state_sequences.items():
+        team = attack.split('-')[0]
+        victim = attack.split('->')[1]
+        team_victim = team + '-' + victim
+        if team_victim not in victim_episodes.keys():
+            victim_episodes[team_victim] = []
+        victim_episodes[team_victim].extend(episodes)
+        victim_episodes[team_victim] = sorted(victim_episodes[team_victim], key=lambda epi: epi[0])  # By start time
+    # Sort by start time across all
+    victim_episodes = {k: v for k, v in sorted(victim_episodes.items(), key=lambda kv: len([epi[0] for epi in kv[1]]))}
+    print('Victims hosts: ', set([team_victim.split('-')[-1] for team_victim in victim_episodes.keys()]))
+
+    attacker_episodes = dict()  # Episodes per (team, attacker)
+    for attack, episodes in state_sequences.items():
+        team = attack.split('-')[0]
+        attacker = (attack.split('->')[0]).split('-')[1]
+        team_attacker = team + '-' + attacker
+        if team_attacker not in attacker_episodes.keys():
+            attacker_episodes[team_attacker] = []
+        attacker_episodes[team_attacker].extend(episodes)
+        attacker_episodes[team_attacker] = sorted(attacker_episodes[team_attacker], key=lambda epi: epi[0])
+    print('Attacker hosts: ', set([team_attacker.split('-')[1] for team_attacker in attacker_episodes.keys()]))
+
+    return attacker_episodes, victim_episodes
+
+
+# Translate technical nodes to human-readable
+def _translate(label, root=False):
     new_label = ""
     parts = label.split("|")
     if root:
-        new_label += 'Victim: '+str(root)+'\n'
+        new_label += 'Victim: ' + str(root) + '\n'
 
     if len(parts) >= 1:
         new_label += verbose_micro[parts[0]]
     if len(parts) >= 2:
-        new_label += "\n"+parts[1]
+        new_label += "\n" + parts[1]
     if len(parts) >= 3:
-        new_label += " | ID: "+parts[2]
+        new_label += " | ID: " + parts[2]
 
     return new_label
-    
-## Per-objective attack graph for dot: 14 Nov (final attack graph) 
-def make_AG(condensed_v_data, condensed_data, state_groups, sev_sinks, datafile, expname):  
```

Comment:
I have extracted some parts into separate methods, as the original one was really huge